blocks/hooks/align.js

```diff
@@ -0,0 +1,144 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+import { assign, includes } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { getWrapperDisplayName } from '@wordpress/element';
+import { addFilter } from '@wordpress/hooks';
+import BlockControls from '../block-controls';
+import BlockAlignmentToolbar from '../block-alignment-toolbar';
+
+/**
+ * Internal dependencies
+ */
+import { getBlockSupport, hasBlockSupport } from '../api';
+
+/**
+ * Filters registered block settings, extending attributes to include `align`.
+ *
+ * @param  {Object} settings Original block settings
+ * @return {Object}          Filtered block settings
+ */
+export function addAttribute( settings ) {
+	if ( hasBlockSupport( settings, 'align' ) ) {
+		// Use Lodash's assign to gracefully handle if attributes are undefined
+		settings.attributes = assign( settings.attributes, {
+			align: {
+				type: 'string',
+			},
+		} );
+	}
+
+	return settings;
+}
+
+/**
+ * Returns an array of valid alignments for a block type depending on its
+ * defined supports. Returns an empty array if block does not support align.
+ *
+ * @param  {String}   blockName Block name to check
+ * @return {String[]}           Valid alignments for block
+ */
+export function getBlockValidAlignments( blockName ) {
+	// Explicitly defined array set of valid alignments
+	const blockAlign = getBlockSupport( blockName, 'align' );
+	if ( Array.isArray( blockAlign ) ) {
+		return blockAlign;
+	}
+
+	const validAlignments = [];
+	if ( true === blockAlign ) {
+		// `true` includes all alignments...
+		validAlignments.push( 'left', 'center', 'right' );
+
+		// ...including wide alignments unless explicitly `false`.
+		if ( hasBlockSupport( blockName, 'wideAlign', true ) ) {
+			validAlignments.push( 'wide', 'full' );
+		}
+	}
+
+	return validAlignments;
+}
+
+/**
+ * Override the default edit UI to include new toolbar controls for block
+ * alignment, if block defines support.
+ *
+ * @param  {Function} BlockEdit Original component
+ * @return {Function}           Wrapped component
+ */
+export function withToolbarControls( BlockEdit ) {
+	const WrappedBlockEdit = ( props ) => {
+		const validAlignments = getBlockValidAlignments( props.name );
+
+		const updateAlignment = ( nextAlign ) => props.setAttributes( { align: nextAlign } );
+
+		return [
+			validAlignments.length > 0 && props.focus && (
+				<BlockControls key="controls">
+					<BlockAlignmentToolbar
```

Comment:
Given that we support the configuration of the supported aligns via an array, we need some way to pass the validAlignments to the BlockAlignmentToolbar. If we set support for align: [ 'center', 'right' ], left align continues to appear although it does not take effect because class and data-align prop is not set. Given that left align is disabled in that case we should not show it.

Reply:
>Given that we support the configuration of the supported aligns via an array, we need some way to pass the validAlignments to the BlockAlignmentToolbar.

An easy enough fix. See 9a2db6e382bb2fe5f7654ba2089f06cfe9a9edff.